util/src/openssl_utils.cpp

```diff
@@ -0,0 +1,258 @@
+
+  ConcordAssertNE(edPkeyCtx, nullptr);
+
+  ConcordAssertEQ(1, EVP_PKEY_keygen_init(edPkeyCtx.get()));
+  ConcordAssertEQ(
+      1, EVP_PKEY_keygen(edPkeyCtx.get(), reinterpret_cast<EVP_PKEY**>(&edPkey)));  // Generate EdDSA key 'edPkey'.
+
+  unsigned char privKey[EdDSASignatureByteSize]{};
+  unsigned char pubKey[EdDSASignatureByteSize]{};
+  size_t privlen{EdDSASignatureByteSize};
+  size_t publen{EdDSASignatureByteSize};
+
+  ConcordAssertEQ(1, EVP_PKEY_get_raw_private_key(edPkey.get(), privKey, &privlen));
+  ConcordAssertEQ(1, EVP_PKEY_get_raw_public_key(edPkey.get(), pubKey, &publen));
+
+  pair<string, string> keyPair;
+  keyPair.first = concordUtils::bufferToHex(reinterpret_cast<const char*>(privKey), (const size_t)privlen, false);
+  keyPair.second = concordUtils::bufferToHex(reinterpret_cast<const char*>(pubKey), (const size_t)publen, false);
+
+  if (KeyFormat::PemFormat == fmt) {
+    keyPair = EdDSAHexToPem(keyPair);
+  }
+  return keyPair;
+}
+
+pair<string, string> OpenSSLCryptoImpl::EdDSAHexToPem(const std::pair<std::string, std::string>& hex_key_pair) const {
+  string privPemString;
+  string pubPemString;
+
+  if (!hex_key_pair.first.empty()) {  // Proceed with private key pem file generation.
+    const auto privKey = concordUtils::unhex(hex_key_pair.first);
+
+    UniquePKEY ed_privKey(EVP_PKEY_new_raw_private_key(NID_ED25519, nullptr, privKey.data(), privKey.size()));
+
+    ConcordAssertNE(nullptr, ed_privKey);
+
+    const char* const tempPrivPem = "/tmp/concordTempPrivKey.pem";
+
+    FILE* fp = fopen(tempPrivPem, "w");
+    PEM_write_PrivateKey(fp, ed_privKey.get(), nullptr, nullptr, 0, nullptr, nullptr);
+    fclose(fp);
+
+    // Read private key from pem file.
+    std::ifstream privPem(tempPrivPem);
+    if (privPem.is_open()) {
+      string temp;
+
+      while (privPem.good()) {
+        getline(privPem, temp);
+        privPemString += temp + '\n';
+      }
+    }
+    remove(tempPrivPem);
+  }
+
+  if (!hex_key_pair.second.empty()) {  // Proceed with public key pem file generation.
+    const auto pubKey = concordUtils::unhex(hex_key_pair.second);
+
+    UniquePKEY ed_pubKey(EVP_PKEY_new_raw_public_key(NID_ED25519, nullptr, pubKey.data(), pubKey.size()));
+
+    ConcordAssertNE(nullptr, ed_pubKey);
+
+    const char* const tempPubPem = "/tmp/concordTempPubKey.pem";
+
+    FILE* fp = fopen(tempPubPem, "w");
+    PEM_write_PUBKEY(fp, ed_pubKey.get());
+    fclose(fp);
+
+    // Read public key from pem file.
+    std::ifstream pubPem(tempPubPem);
+    if (pubPem.is_open()) {
+      string temp;
+
+      while (pubPem.good()) {
+        getline(pubPem, temp);
+        pubPemString += temp + '\n';
+      }
+    }
+    remove(tempPubPem);
+  }
+  return make_pair(privPemString, pubPemString);
+}
+
+KeyFormat OpenSSLCryptoImpl::getFormat(const std::string& key) const {
+  return (key.find("BEGIN") != std::string::npos) ? KeyFormat::PemFormat : KeyFormat::HexaDecimalStrippedFormat;
```

Comment:
This is not a robust way of figuring out the format of the file, but a hack which works for our specific use case only and might break when adding additional formats.
You should find a more robust way of figuring out the format (example: dont use raw strings, propagate the information)

Reply:
This way of checking is an existing way of checking.

Reply:
TODO added 🙂